app/Http/Controllers/ThirdParty/CallPowerController.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+namespace Chompy\Http\Controllers\ThirdParty;
+
+use Exception;
+use Illuminate\Http\Request;
+use Chompy\Services\Rogue;
+use Chompy\Http\Controllers\Controller;
+
+class CallPowerController extends Controller
+{
+
+    /**
+     * Create a controller instance.
+     *
+     * @return void
+     */
+    public function __construct(Rogue $rogue)
+    {
+    	$this->rogue = $rogue;
+
+    }
+
+    /**
+     *
+     * @param Request $request
+     */
+    public function store(Request $request)
+    {
+        $request->validate([
+            'mobile' => 'required',
+            'callpower_campaign_id' => 'required|integer',
+            'status' => 'required|string',
+            'call_timestamp' => 'required|date',
+            'call_duration' => 'required|integer',
+            'campaign_target_name' => 'required|string',
+            'campaign_target_title' => 'required|string',
+            'campaign_target_district' => 'required|string',
+            'callpower_campaign_name' => 'required|string',
+            'number_dialed_into' => 'required',
+        ]);
+
+        // Using the mobile number, get or create a northstar_id.
+        $user = $this->getOrCreateUser($request['mobile']);
```

Comment:
This should all be happening in a queued job, like we do for other imports.

Reply:
@DFurnes I don't think that Chompy does this? For RTV, seems like it does the same thing as it does [here](https://app.ghostinspector.com/tests/5c49f1dd638e692a231ece5f)?

Reply:
I think you might have pasted the wrong link in there?

This is how our other imports in Chompy work, for example, here's where [we kick off queued jobs for RTV imports in `ImportsController`](https://github.com/DoSomething/chompy/blob/a08a67897c1835846cae30667b9e1d6f7e5eff26/app/Http/Controllers/ImportController.php#L62-L75). We do this because otherwise the whole "import" has to happen before Chompy can render a response, and we'd run into request timeouts for anything more than the shortest of CSVs.

This endpoint here, admittedly, is not parsing a whole CSV! But we still don't want _all_ this work (let's say, 200-1000ms to get or create the Northstar user, and 500-1000ms to create the post) to take up time on the main process. Chompy has a limited number of requests it can handle at once, and so if we have really "hefty" requests like this, we'd potentially use up all our resources on something that we don't need a response from right away.

By putting this work into a queued job, we also automatically gain retries for "free" (so if, for example, we find out that CallPower was sending data that caused Northstar or Rogue to error out, we could fix the bug in our systems and simply retry the failing job, rather than having to ask Josh to re-trigger the event on their end).

Reply:
oh I see what you're saying now, thank you for this amazing description @DFurnes ! 

So this should ALL live in a job, not just the `getOrCreateUser` function?

Reply:
> So this should ALL live in a job, not just the `getOrCreateUser` function?

Yes! You can use [`ImportController`](https://github.com/DoSomething/chompy/blob/a08a67897c1835846cae30667b9e1d6f7e5eff26/app/Http/Controllers/ImportController.php) and the [`CreateRockTheVotePostsInRogue`](https://github.com/DoSomething/chompy/blob/a08a67897c1835846cae30667b9e1d6f7e5eff26/app/Jobs/CreateRockTheVotePostInRogue.php) files as an example for how this should be broken up.

Reply:
awesome, this makes so much sense, thank you again, @DFurnes ! This is updated!